components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMServiceTest.java

```diff
@@ -1054,4 +1058,89 @@ private void startTenantFlow() {
+        validCallbackList.add("https://wso2.com?dummy1=1&dummy=2");
+        List<String> invalidCallbackList = new ArrayList<>();
+        invalidCallbackList.add("https://wso2.com/");
+        invalidCallbackList.add("https://wso2.com/?dummy1");
+        invalidCallbackList.add("https://wso2.com/?dummy1=1&dummy=2");
+        return new Object[][]{
+                {redirectUriList, validCallbackList, invalidCallbackList}
+        };
+    }
+
+    @Test(dataProvider = "redirectUriWithQueryParamsProvider")
+    public void registerApplicationTestWithRedirectURls(List<String> redirectUri, List<String> validCallbackList,
+                                                        List<String> invalidCallbackList) throws Exception {
+
+        mockApplicationManagementService = mock(ApplicationManagementService.class);
+
+        Whitebox.setInternalState(dcrmService, "oAuthAdminService", mockOAuthAdminService);
+
+        startTenantFlow();
+
+        dummyGrantTypes.add("implicit");
+        applicationRegistrationRequest.setGrantTypes(dummyGrantTypes);
+
+        String grantType = StringUtils.join(applicationRegistrationRequest.getGrantTypes(), " ");
+
+        ServiceProvider serviceProvider = new ServiceProvider();
+
+        DCRDataHolder dcrDataHolder = DCRDataHolder.getInstance();
+        dcrDataHolder.setApplicationManagementService(mockApplicationManagementService);
+        when(mockApplicationManagementService.getServiceProvider(dummyClientName, dummyTenantDomain)).thenReturn
+                (null, serviceProvider);
+
+        applicationRegistrationRequest.setRedirectUris(redirectUri);
+
+        OAuthConsumerAppDTO oAuthConsumerApp = new OAuthConsumerAppDTO();
+        oAuthConsumerApp.setApplicationName(dummyClientName);
+
+        oAuthConsumerApp.setGrantTypes(grantType);
+        oAuthConsumerApp.setOAuthVersion(OAUTH_VERSION);
+        oAuthConsumerApp.setCallbackUrl(OAuthConstants.CALLBACK_URL_REGEXP_PREFIX + createRegexPattern(redirectUri));
+
+        when(mockOAuthAdminService
+                .getOAuthApplicationDataByAppName(dummyClientName)).thenReturn(oAuthConsumerApp);
+        when(mockOAuthAdminService.registerAndRetrieveOAuthApplicationData(any(OAuthConsumerAppDTO.class)))
+                .thenReturn(oAuthConsumerApp);
+
+        Application application = dcrmService.registerApplication(applicationRegistrationRequest);
+        assertEquals(application.getClientName(), dummyClientName);
+        String regexp = application.getRedirectUris().get(0)
+                .substring(OAuthConstants.CALLBACK_URL_REGEXP_PREFIX.length());
+        for (String validCallback : validCallbackList) {
+            assertTrue(validCallback.matches(regexp));
+        }
+        for (String invalidCallback : invalidCallbackList) {
+            assertFalse(invalidCallback.matches(regexp));
+        }
+    }
+
+    private String createRegexPattern(List<String> redirectURIs) throws DCRMException {
```

Comment:
This is the same method from the service it seems. Can you make the service method protected and re-use here?

Reply:
Done